Unexpected restarts of the coder pod

Hello, I see sporadically this error on a GKE cluster:

(coder 0.11.0, GKE 1.24.6)

```
2022-10-31 07:34:58.861 [INFO]	(coderd.tailnet)	<./tailnet/conn.go:200>	NewConn.func3	wireguard status	{"status": "\u0026{AsOf:2022-10-31 07:34:58.861635364 +0000 UTC m=+1406.588420654 Peers:[{TxBytes:802036 RxBytes:5117168 LastHandshake:2022-10-31 07:33:55.787139294 +0000 UTC NodeKey:nodekey:cd32e74502524f98a828bf4adb7d4d23b0dd8ad0f0ce25006bf49c01f3e89123}] LocalAddrs:[{Addr:xxx.xxx.xxx.xxx:58148 Type:stun} {Addr:10.108.2.51:58148 Type:local}] DERPs:1}", "err": null}
2022-10-31 07:34:58.861 [INFO]	(coderd.tailnet)	<./tailnet/conn.go:456>	(*Conn).sendNode.func1	sending node	{"node": {"id": 5166511059428960592, "as_of": "2022-10-31T07:34:58.861635364Z", "key": "nodekey:33a3060e0d5648c2e31468634fc7811d4abb089049905814aac0eabd4f6d0c40", "disco": "discokey:282dfd24191766d8cc5d93b40231dced685b3227020aed41b18f57cc2eeaaf0e", "preferred_derp": 999, "derp_latency": {"999-v4": 0.008294595}, "addresses": ["fd7a:115c:a1e0:4970:805a:f00a:6f25:63b/128"], "allowed_ips": ["fd7a:115c:a1e0:4970:805a:f00a:6f25:63b/128"], "endpoints": ["xxx.xxx.xxx.xxx:58148", "xxx.xxx.xxx.xxx:58148"]}}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x683178]

goroutine 107607 [running]:
bufio.(*Writer).Flush(0x800?)
	/opt/hostedtoolcache/go/1.19.2/x64/src/bufio/bufio.go:623 +0x18
net/http.(*chunkWriter).flush(0xc0150dc840?)
	/opt/hostedtoolcache/go/1.19.2/x64/src/net/http/server.go:402 +0x3f
net/http.(*response).Flush(0xc01bbaa000)
	/opt/hostedtoolcache/go/1.19.2/x64/src/net/http/server.go:1707 +0x48
github.com/coder/coder/coderd/tracing.(*StatusWriter).Flush(0x7f1929df9e78?)
	/home/runner/work/coder/coder/coderd/tracing/status_writer.go:81 +0x35
github.com/coder/coder/coderd/httpapi.ServerSentEventSender.func1()
	/home/runner/work/coder/coder/coderd/httpapi/httpapi.go:206 +0x202
created by github.com/coder/coder/coderd/httpapi.ServerSentEventSender
	/home/runner/work/coder/coder/coderd/httpapi/httpapi.go:191 +0x37d
```

--- a/coderd/httpapi/httpapi.go
+++ b/coderd/httpapi/httpapi.go
@@ -245,7 +245,7 @@ func ServerSentEventSender(rw http.ResponseWriter, r *http.Request) (sendEvent f
 
 		event := sseEvent{
 			payload: buf.Bytes(),
-			errC:    make(chan error, 1),
+			errC:    make(chan error, 1), // Buffered to prevent deadlock.
 		}
 
 		select {
@@ -253,10 +253,20 @@ func ServerSentEventSender(rw http.ResponseWriter, r *http.Request) (sendEvent f
 			return r.Context().Err()
 		case <-ctx.Done():
 			return ctx.Err()
-		case eventC <- event:
-			return <-event.errC
 		case <-closed:
 			return xerrors.New("server sent event sender closed")
+		case eventC <- event:
+			// Re-check closure signals after sending the event to allow
+			// for early exit. We don't check closed here because it
+			// can't happen while processing the event.
+			select {
+			case <-r.Context().Done():
+				return r.Context().Err()
+			case <-ctx.Done():
+				return ctx.Err()
+			case err := <-event.errC:
+				return err
+			}
 		}
 	}
 

--- a/coderd/httpapi/httpapi.go
+++ b/coderd/httpapi/httpapi.go
@@ -6,14 +6,13 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
-	"io"
 	"net/http"
 	"reflect"
 	"strings"
-	"sync"
 	"time"
 
 	"github.com/go-playground/validator/v10"
+	"golang.org/x/xerrors"
 
 	"github.com/coder/coder/coderd/tracing"
 	"github.com/coder/coder/codersdk"
@@ -174,8 +173,7 @@ func WebsocketCloseSprintf(format string, vars ...any) string {
 	return msg
 }
 
-func ServerSentEventSender(rw http.ResponseWriter, r *http.Request) (func(ctx context.Context, sse codersdk.ServerSentEvent) error, error) {
-	var mu sync.Mutex
+func ServerSentEventSender(rw http.ResponseWriter, r *http.Request) (sendEvent func(ctx context.Context, sse codersdk.ServerSentEvent) error, closed chan struct{}, err error) {
 	h := rw.Header()
 	h.Set("Content-Type", "text/event-stream")
 	h.Set("Cache-Control", "no-cache")
@@ -187,37 +185,50 @@ func ServerSentEventSender(rw http.ResponseWriter, r *http.Request) (func(ctx co
 		panic("http.ResponseWriter is not http.Flusher")
 	}
 
-	// Send a heartbeat every 15 seconds to avoid the connection being killed.
+	closed = make(chan struct{})
+	type sseEvent struct {
+		payload []byte
+		errC    chan error
+	}
+	eventC := make(chan sseEvent)
+
+	// Synchronized handling of events (no guarantee of order).
 	go func() {
+		defer close(closed)
+
+		// Send a heartbeat every 15 seconds to avoid the connection being killed.
 		ticker := time.NewTicker(time.Second * 15)
 		defer ticker.Stop()
 
 		for {
+			var event sseEvent
+
 			select {
 			case <-r.Context().Done():
 				return
+			case event = <-eventC:
 			case <-ticker.C:
-				mu.Lock()
-				_, err := io.WriteString(rw, fmt.Sprintf("event: %s\n\n", codersdk.ServerSentEventTypePing))
-				if err != nil {
-					mu.Unlock()
-					return
+				event = sseEvent{
+					payload: []byte(fmt.Sprintf("event: %s\n\n", codersdk.ServerSentEventTypePing)),
 				}
-				f.Flush()
-				mu.Unlock()
 			}
+
+			_, err := rw.Write(event.payload)
+			if event.errC != nil {
+				event.errC <- err
+			}
+			if err != nil {
+				return
+			}
+			f.Flush()
 		}
 	}()
 
-	sendEvent := func(ctx context.Context, sse codersdk.ServerSentEvent) error {
-		if ctx.Err() != nil {
-			return ctx.Err()
-		}
-
+	sendEvent = func(ctx context.Context, sse codersdk.ServerSentEvent) error {
 		buf := &bytes.Buffer{}
 		enc := json.NewEncoder(buf)
 
-		_, err := buf.Write([]byte(fmt.Sprintf("event: %s\ndata: ", sse.Type)))
+		_, err := buf.WriteString(fmt.Sprintf("event: %s\ndata: ", sse.Type))
 		if err != nil {
 			return err
 		}
@@ -232,16 +243,22 @@ func ServerSentEventSender(rw http.ResponseWriter, r *http.Request) (func(ctx co
 			return err
 		}
 
-		mu.Lock()
-		defer mu.Unlock()
-		_, err = rw.Write(buf.Bytes())
-		if err != nil {
-			return err
+		event := sseEvent{
+			payload: buf.Bytes(),
+			errC:    make(chan error, 1),
 		}
-		f.Flush()
 
-		return nil
+		select {
+		case <-r.Context().Done():
+			return r.Context().Err()
+		case <-ctx.Done():
+			return ctx.Err()
+		case eventC <- event:
+			return <-event.errC
+		case <-closed:
+			return xerrors.New("server sent event sender closed")
+		}
 	}
 
-	return sendEvent, nil
+	return sendEvent, closed, nil
 }

--- a/coderd/workspaces.go
+++ b/coderd/workspaces.go
@@ -867,7 +867,7 @@ func (api *API) watchWorkspace(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	sendEvent, err := httpapi.ServerSentEventSender(rw, r)
+	sendEvent, senderClosed, err := httpapi.ServerSentEventSender(rw, r)
 	if err != nil {
 		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
 			Message: "Internal error setting up server-sent events.",
@@ -875,6 +875,10 @@ func (api *API) watchWorkspace(rw http.ResponseWriter, r *http.Request) {
 		})
 		return
 	}
+	// Prevent handler from returning until the sender is closed.
+	defer func() {
+		<-senderClosed
+	}()
 
 	// Ignore all trace spans after this, they're not too useful.
 	ctx = trace.ContextWithSpan(ctx, tracing.NoopSpan)
@@ -884,6 +888,8 @@ func (api *API) watchWorkspace(rw http.ResponseWriter, r *http.Request) {
 	for {
 		select {
 		case <-ctx.Done():
+			return
+		case <-senderClosed:
 			return
 		case <-t.C:
 			workspace, err := api.Database.GetWorkspaceByID(ctx, workspace.ID)

